currencyrate/currencyrate.py

```diff
@@ -0,0 +1,148 @@
+    Source('coindesk',
+           'https://api.coindesk.com/v1/bpi/currentprice/{currency}.json',
+           ['bpi', '{currency}', 'rate_float']),
+    # e.g. {"data":{"base":"BTC","currency":"USD","amount":"19414.63"}}
+    Source('coinbase',
+           'https://api.coinbase.com/v2/prices/spot?currency={currency}',
+           ['data', 'amount']),
+    # e.g. {  "USD" : {"15m" : 6650.3, "last" : 6650.3, "buy" : 6650.3, "sell" : 6650.3, "symbol" : "$"},  "AUD" : {"15m" : 10857.19, "last" : 10857.19, "buy" : 10857.19, "sell" : 10857.19, "symbol" : "$"},...
+    Source('blockchain.info',
+           'https://blockchain.info/ticker',
+           ['{currency}', 'last']),
+]
+
+
+def get_currencyrate(plugin, currency, req_template, response_members):
+    # NOTE: Bitstamp has a DNS/Proxy issues that can return 404
+    # Workaround: retry up to 5 times with a delay
+    currency_lc = currency.lower()
+    url = req_template.format(currency_lc=currency_lc, currency=currency)
+    for _ in range(5):
+        r = requests.get(url, proxies=plugin.proxies)
+        if r.status_code != 200:
+            time.sleep(1)
+            continue
+        break
+
+    if r.status_code != 200:
+        plugin.log(level='info', message='{}: bad response {}'.format(url, r.status_code))
+        return None
+
+    json = r.json()
+    for m in response_members:
+        expanded = m.format(currency_lc=currency_lc, currency=currency)
+        if expanded not in json:
+            plugin.log(level='debug', message='{}: {} not in {}'.format(url, expanded, json))
+            return None
+        json = json[expanded]
+
+    try:
+        return Millisatoshi(int(10**11 / float(json)))
+    except Exception:
+        plugin.log(level='info', message='{}: could not convert {} to msat'.format(url, json))
+        return None
+
+
+def set_proxies(plugin):
+    config = plugin.rpc.listconfigs()
+    if 'always-use-proxy' in config and config['always-use-proxy']:
+        paddr = config['proxy']
+        # Default port in 9050
+        if ':' not in paddr:
+            paddr += ':9050'
+        plugin.proxies = {'https': 'socks5h://' + paddr,
+                          'http': 'socks5h://' + paddr}
+    else:
+        plugin.proxies = None
+
+
+# Don't grab these more than once per hour.                        
+@cached(cache=TTLCache(maxsize=1024, ttl=3600))
+def get_rates(plugin, currency):
+    rates = {}
+    for s in sources:
+        r = get_currencyrate(plugin, currency, s.urlformat, s.replymembers)
+        if r is not None:
+            rates[s.name] = r
+
+    return rates
+
```

Comment:
We could have a manual cache update method/trigger incase someone wants/needs to have fresh rates urgently.

```suggestion
@plugin.method("currencyupdate")
def currencyupdate(plugin)
    ...
```

Reply:
Or maybe have an optional parameter on `currencyconvert` and `currencyrates` that forces a (blocking) cache update.

Reply:
Or maybe have an optional parameter on `currencyrates` and `currencyconvert` that sets the caches to e.g. `ttl=300` temporarily just for that call... if something like that is possible.

Reply:
<shrug> Probably, but can we add it later when we figure it out?